### PR TITLE
codeintel: Add configuration to backend

### PR DESCRIPTION
**Pull request by efritz**
This PR adds a database table, a set of dbstore methods to manipulate that table, and a graphql layer that allows site-admins to set data retention policies and index scheduling rules per-repository and globally.

A UI to match is coming soon - please pay particular attention to the graphql schema changes as the major discussion item.

### CODE REVIEW

File: migrations/frontend/1528395868_add_lsif_configuration_policy.up.sql [hunk 1]
```diff
@@ -0,0 +1,27 @@
+BEGIN;
+
+CREATE TABLE lsif_configuration_policies (
+    id SERIAL PRIMARY KEY,
+    repository_id int,
```
Comment (Strum355) [inline migrations/frontend/1528395868_add_lsif_configuration_policy.up.sql]: Do we care to make this a foreign key? Im not sure on how deleted repositories are handled (soft delete only?)

**Review by Strum355 — APPROVED** [review]

File: enterprise/internal/codeintel/stores/dbstore/configuration_policies.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+// scanFirstConfigurationPolicy scans a slice of configuration policies from the return value of `*Store.query`
+// and returns the first.
+func scanFirstConfigurationPolicy(rows *sql.Rows, err error) (ConfigurationPolicy, bool, error) {
+	scanConfigurationPolicies, err := scanConfigurationPolicies(rows, err)
+	if err != nil || len(scanConfigurationPolicies) == 0 {
+		return ConfigurationPolicy{}, false, err
+	}
+	return scanConfigurationPolicies[0], true, nil
```
Comment (tjdevries) [inline enterprise/internal/codeintel/stores/dbstore/configuration_policies.go:73]: is there any chance this just loads a ton of rows for no reason?

Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/configuration_policies.go:73]: The convention I'm following here is:

- have scanXes that scans a list of Xes
- have scanFirstX that scans a list of Xes then takes the first

The understanding is that you only use the second type of method with queries that have a limit or a natural condition that allows only allows up to one result row (a count query, something with WHERE id = %s and id is a unique value, etc).

It's _possible_, but it's not used in that way.

File: migrations/frontend/1528395868_add_lsif_configuration_policy.up.sql (hunk 1 above)
Comment (efritz) [reply migrations/frontend/1528395868_add_lsif_configuration_policy.up.sql]: We could make it a foreign key, but I've been purposefully been not hard-linking data from code intel features to the other features (only having value references).

This should make it easier when we inevitably need to split code intel into another database, and all of our foreign keys become cross-db keys.

Comment (efritz) [reply migrations/frontend/1528395868_add_lsif_configuration_policy.up.sql]: To answer the other part: repositories are soft-deleted, and we have a janitor task that will remove _all_ data from a deleted repository after some timeout, so data retention does not need to apply here anymore.

We could delete these rows or leave them. We could delete them in the janitor after a timeout as well.

File: enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/configuration_policy.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	case "GIT_COMMIT":
+		return gql.GitObjectTypeCommit
+	case "GIT_TAG":
+		return gql.GitObjectTypeTag
+	case "GIT_TREE":
+		return gql.GitObjectTypeTree
+	default:
+	}
+
+	return ""
```
Comment (tjdevries) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/configuration_policy.go:41]: is it weird to have the return outside of the switch statement if the default case does nothing?
Also, shouldn't it error if it's something else?

Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/configuration_policy.go:41]: I generally try to use the default as a fallthrough for stylistic reasons (I don't have any better reasons to list here).

Yeah we could error instead here just to catch future changes in what we accept but not what we display.

File: enterprise/internal/codeintel/stores/dbstore/configuration_policies.go (hunk 2 above)
Comment (tjdevries) [reply enterprise/internal/codeintel/stores/dbstore/configuration_policies.go:73]: Got it, makes sense to me

**Pull request merged**